src/Microsoft.Android.Build.BaseTasks/AsyncTask.cs

```diff
@@ -0,0 +1,431 @@
+// https://github.com/xamarin/Xamarin.Build.AsyncTask/blob/db4ce14dacfef47435c238b1b681c124e60ea1a0/Xamarin.Build.AsyncTask/AsyncTask.cs
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Utilities;
+using Microsoft.Build.Framework;
+using System.Threading;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Android.Build.Tasks
+{
+	/// <summary>
+	/// Base class for tasks that need long-running cancellable asynchronous tasks 
+	/// that don't block the UI thread in the IDE.
+	/// </summary>
+	public class AsyncTask : Task, ICancelableTask
```

Comment:
Since this is new API -- the namespace changed -- we should revisit some API design decisions:

 1. This class isn't *really* usable by itself; nobody will "reasonably" do `new AsyncTask()`.  As such, should we make this type `abstract` and make the constructor `protected`?
 2. Should we "flatten" `AsyncTask` and `AndroidAsyncTask` / "Task-ify `AsyncTask`, by way of moving `RunTaskAsync()` and `RunTask()` to this type?  (Leave `TaskPrefix` in `AndroidAsyncTask`.)
 3. Variation on (2): it's not at all obvious to subclasses what they should *do*.  It declares no new `virtual` or `abstract` methods, meaning `Execute()` is the only obvious method to override.  (2) could help solve that.

Is there any other API cleanup we should consider?

Reply:
I'm in favour of making it abstract and keeping the existing tooling. Because a custom task generally involves subclassing and writting code I would not expect anyone to try using this class directly. 

When this was a generic class to be used anywhere it made sense to have our own `AndroidAsyncTask` to make things easier for us. As that is no longer required we can probably folder the AndroidAsyncTask into this one. The only down side on that is the number of changes that will be needed when we bump.

Reply:
I like the idea of moving some things into `AsyncTask`, while keeping `AndroidAsyncTask` around to reduce the number of changes required. Though, we could also remove `AndroidAsyncTask`, and merge everything into a single `AsyncTask` (or continue using the `AndroidAsyncTask` name) if that would be preferable?

Reply:
I kept these as two separate classes for now, while moving most functionality into the now abstract AsyncTask.

Reply:
`AndroidAsyncTask` and `AsyncTask` have been merged into a single `AsyncTask`.